### PR TITLE
Merge mit pdos [5] (related to #261)

### PR DESCRIPTION
**Pull request by kimjungwow**
### Rust 코드에 반영되었는지 한 번 더 확인 후 리뷰 요청 드리겠습니다 (related to #261 )

- [X] https://github.com/mit-pdos/xv6-riscv/commit/a93321cb2547dbb48bf8ce9ad623ac19eefbecea : test pointer checking in copyin, copyout, copyinstr
  - copyin1, copyin2, copyout1, copyout2, copyinstr1, copyinstr2 usertest 추가 -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/e3b7058907dff779cf94e23bf6bb84245faf481d : streamline copyin/copyout code in usertests
  - read/write 시 에러인 경우 return value 때문에 생기는 에러 고침
  - `console.rs, fs.rs` 이미 반영됨
  - `pipe.rs` : `PipeError::InvalidCopyin`를 `PipeError::InvalidCopyin(usize)`로 변경
  - usertest 추가
- [X] https://github.com/mit-pdos/xv6-riscv/commit/76d6c57edec14071156e7780f7e8e08c200cf166 : test copyinstr()'s handling of the terminating null
  - usertest 추가 -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/d8fe1773b26758c7c7b8f36724cd822555b33612 : test string system call arguments that cross over the end of the last…
  - usertest 추가 -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/c31d35d8031c88b7e1ea8657cc9806dfdd4c3ef9 : Merge branch 'riscv' into riscv
  - 머지 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/6b593b9eaef575d7d922c202ca5e4d3a0ff317bf : Merge pull request #4 from 4ge32/riscv
  - 머지 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/d28c5d9837a50b6ae26f03bc84a9dc0d7792562e :  Merge pull request #6 from jrrk2/riscv
  - 머지 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/12c131de3eb87e725a6a4a385fcedf6b78f9b510 : Merge pull request #8 from rui314/cat-fix
  - 머지 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/93ca5550b8888bd76815b75203fcea02ef295aa4 : Merge pull request #10 from takahirox/VirtioInterruptAck
  - 머지 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/050234233521bfe62da9bd5f93b0e78de30d810b : Merge pull request #9 from d0iasm/riscv
  - `proc.c`의 `initcode[]` 변경된 것 `proc.rs`에서 재확인 
- [X] https://github.com/mit-pdos/xv6-riscv/commit/96a20b0c0fc8c0376e02f261361e22d6c6c839f9 : Generate initcode without compressed extensions
  - `Makefile`만 변경 -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/ffcb883adff508497f71fae6ca72a8fa70483edc : Patch : Fix minor typos
  - 이전의 다른 commit에서 이미 반영된 내용 (C, Rust 모두)
- [X] https://github.com/mit-pdos/xv6-riscv/commit/b557e7c32e935b9bb2f5d8ed8503de52f43cf87f : Update ramdisk.c
  - `ramdisk.c` -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/9eff4b0871ca68376bc409c991fbce414e700921 : Write interrupt ack register in virtio_disk_intr()
  - 이전의 다른 commit에서 이미 반영된 내용 (C, Rust 모두)
- [X] https://github.com/mit-pdos/xv6-riscv/commit/c24844714bf68feb8965c16be6ad9c82e27cc530 : update initcode to avoid using the compressed extension
  - 이전의 다른 commit에서 이미 반영된 내용 (C, Rust 모두)
- [X] https://github.com/mit-pdos/xv6-riscv/commit/f14aa421c456875fbb5fff0ef2f5c4154f11e38d : Generate initcode without compressed extensions
    - 이전의 다른 commit에서 이미 반영된 내용 (`Makefile`)  
- [X] https://github.com/mit-pdos/xv6-riscv/commit/2db95f08b14f2779a31cb9e3e830f762f36f2660 : Typo (thanks yt <jaichenhengjie@qq.com>)
  - `entry.S` 오타 수정 -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/737bd3a55d9380906f097f1d82d47fe2b23bf486 : Ack bug finders.
  - `README` 오타 수정 -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/5dace6737104d817674cfc41d41719ba58009df6 : Typo (thanks yt <jaichenhengjie@qq.com>)
  - `entry.S` 오타 수정 -> **Rust 해당 사항 없음**  
- [X] https://github.com/mit-pdos/xv6-riscv/commit/fa71d7fca2914af0a22f4cae3f40c6346b5fdc2c : Ack bug finders.
  - `README` 오타 수정 -> **Rust 해당 사항 없음** 
- [X] https://github.com/mit-pdos/xv6-riscv/commit/84dcf8a738f5349fdb37449502086b2b3c95763c : copyright and 6.S081
  - `README` 오타 수정 -> **Rust 해당 사항 없음**   
- [X] https://github.com/mit-pdos/xv6-riscv/commit/468946781fd551d69252f13a41ffc77c5e30fae9 : copyright and 6.S081
  - `README` 오타 수정 -> **Rust 해당 사항 없음**     
- [X] https://github.com/mit-pdos/xv6-riscv/commit/315c37c0f1e7c05f7b7d0ec145a2002b0612c49c : Update to the latest specs
  - pdf 파일들 변경 -> **Rust 해당 사항 없음**
- [X] https://github.com/mit-pdos/xv6-riscv/commit/56e7cebdb957ef12fcd9c0800887cd44fb5c250d: Update to the latest specs
  - pdf 파일들 변경 -> **Rust 해당 사항 없음**

**위 까지가 2020/8/11**

- [X] https://github.com/mit-pdos/xv6-riscv/commit/d32eac366fd06ec58cc3b34a3ca1824864bd7618 : pa isn't needed in the uvmunmap() loop
  - `vm.c:uvnunmap()`에서 `pa`를 if(do_free) 내부에서 선언하기 -> `vm.rs`에 적용 완료
- [X] https://github.com/mit-pdos/xv6-riscv/commit/e1a2cd59faf0cf7a3b315cefff11d68b68ec54a7 : Merge branch 'riscv' of g.csail.mit.edu:xv6-dev into riscv
  - 이전 커밋들에서 이미 적용됨
- [X] https://github.com/mit-pdos/xv6-riscv/commit/70c6fe861e64195032083a9fd59a81bd113d1b0b : modify uvmunmap() to be in aligned pages
  - fix a bug in fork() recovering from out of memory
  - `proc.rs, vm.rs` 반영 완료
- [X] https://github.com/mit-pdos/xv6-riscv/commit/4c22c54480f020c36de120ce868912c022e48113 : try to handle a few of the possible out-of-memory errors in fork()
  - `proc.c:proc_pagetable()`에서 메모리 할당이 실패한 경우 `uvmfree(), uvmunmap()`을 적절히 실행 : `proc.rs:proc_pagetable()`에 적용 완료
  - `proc.rs:alloc()`에서 `proc_pagetable()` 리턴 값이 0인 경우를 따로 처리하기 때문에, 리턴 타입을 Result로 변경함
  - `proc.c:allocproc()`에서 `proc_pagetable()` 호출했는데 0이 리턴된 경우, `freeproc()` 실행 후 0 리턴하기 : `proc.rs:alloc()`에 적용 완료
- [X] https://github.com/mit-pdos/xv6-riscv/commit/adee82c3e7334a09996c0fe9cc75d9a80abc81c8 : handle another out-of-memory in fork(). the policy here is not consis…
  - handle another out-of-memory in fork(). the policy here is not consistent, since other calls (e.g. exec()) panic on out of memory.
  - `vm.c`의 `uvmcreate()`에서 `kalloc()` 실패 시 panic이 아닌 0을 리턴 : `vm.rs`의 `alloc_root()`의 리턴 타입을 `Result<(), ()>`로 변경 완료

  **위 까지가 2020/8/13**  

- [X] https://github.com/mit-pdos/xv6-riscv/commit/4a87a0ae8bc4a186842fb5f57454412de56f52bc : fix a bug in the out-of-memory handling code in exec
  - `exec.c::exec()`에서 `uvmalloc()`이 실패한 경우 `sz`를 변경하지 않도록 함 : `exec.rs::exec()`에 `sz1` 추가
  - `vm.c::uvmalloc()`에서 `a`를 for loop에서 선언 : Rust 해당 사항 없음

  **위 까지가 2020/8/14**  

- [X] https://github.com/mit-pdos/xv6-riscv/commit/b154cf091c57db1b412a022490c0cec009b4cf6b : x
  - `entry.S, user/init.c, user/initcode.S` : Rust 해당 사항 없음

  **위 까지가 2020/8/15**  

- [X] https://github.com/mit-pdos/xv6-riscv/commit/e283514ad038b7a14f58413e809c7a74766b738e : pa isn't needed in the uvmunmap() loop
  - 리베이스 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/a8dd44cf595bb691ec21a16e01eaaf9b527a05ca : modify uvmunmap() to be in aligned pages
  - 리베이스 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/bc51626aab45f420652f20948a1ffa5398ad9446 : try to handle a few of the possible out-of-memory errors in fork()
  - 리베이스 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/8dc71a257d80d1e2f56cd7f931ca389fbb727b5c : handle another out-of-memory in fork(). the policy here is not consis…
  - 리베이스 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/f2ec6777bd530f949c235d7f0386286314a2f601 : fix a bug in the out-of-memory handling code in exec
  - 리베이스 커밋 : 이미 수정된 내용이므로 반영 X
- [X] https://github.com/mit-pdos/xv6-riscv/commit/95dad4c061d6e1f036d75cc041367a8f12b92982 : x
  - 리베이스 커밋 : 이미 수정된 내용이므로 반영 X

  **위 까지가 2020/8/17**

### CODE REVIEW

**Comment by kimjungwow** [comment]
@jeehoonkang 교수님, 리뷰 요청드립니다.

- 2020/8/11 ~ 2020/8/17 의 커밋들입니다.
- xv6/mit-pdos repo에도 계속 커밋이 올라오고 있어서.. **2020/11/25 기준으로 #261 은 약 60% 진행했습니다.**

File: kernel-rs/src/proc.rs [hunk 1]
```diff
@@ -582,7 +582,10 @@ impl ProcessSystem {
                 data.trapframe = page.into_usize() as *mut Trapframe;
 
                 // An empty user page table.
-                data.pagetable = proc_pagetable(p as *const _ as *mut _);
+                data.pagetable = ok_or!(proc_pagetable(p as *const _ as *mut _), {
+                    freeproc(guard);
```
Comment (jeehoonkang) [inline kernel-rs/src/proc.rs:586]: 이게 약간 의심스러운데... 혹시 어떤 일을 하는지 간단히 설명 부탁드려요.

Comment (jeehoonkang) [reply kernel-rs/src/proc.rs:586]: 의심스럽다기보다는 dtor/drop을 잘 활용해서 refactoring해야할 것 같다는 느낌입니다.  아래에서 err 리턴하는 다른 케이스도 마찬가지

File: kernel-rs/src/proc.rs [hunk 2]
```diff
@@ -893,52 +896,59 @@ unsafe fn freeproc(mut p: ProcGuard) {
...
     pagetable.uvmfree(sz);
 }
 
 /// A user program that calls exec("/init").
 /// od -t xC initcode
-const INITCODE: [u8; 51] = [
-    0x17, 0x05, 0, 0, 0x13, 0x05, 0x05, 0x02, 0x97, 0x05, 0, 0, 0x93, 0x85, 0x05, 0x02, 0x9d, 0x48,
-    0x73, 0, 0, 0, 0x89, 0x48, 0x73, 0, 0, 0, 0xef, 0xf0, 0xbf, 0xff, 0x2f, 0x69, 0x6e, 0x69, 0x74,
-    0, 0, 0x01, 0x20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+const INITCODE: [u8; 52] = [
```
Comment (jeehoonkang) [inline kernel-rs/src/proc.rs:948]: 혹시 이제 qemu 최신 버전에서도 잘 돌아가나요?

File: kernel-rs/src/vm.rs [hunk 3]
```diff
@@ -236,10 +235,11 @@ impl<A: VAddr> PageTable<A> {
         }
     }
 
-    pub fn alloc_root(&mut self) {
-        let mut page = unsafe { kernel().alloc() }.expect("PageTable new: out of memory");
+    pub fn alloc_root(&mut self) -> Result<(), ()> {
+        let mut page = unsafe { some_or!(kernel().alloc(), return Err(())) };
```
Comment (jeehoonkang) [inline kernel-rs/src/vm.rs:239]: some_or! 대신에 https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or

File: kernel-rs/src/vm.rs [hunk 4]
```diff
@@ -453,35 +453,32 @@ impl PageTable<UVAddr> {
...
-        let mut a = pgrounddown(va.into_usize());
-        let last = pgrounddown(va.into_usize() + size - 1usize);
+    /// Remove npages of mappings starting from va. va must be
+    /// page-aligned. The mappings must exist.
+    /// Optionally free the physical memory.
+    pub unsafe fn uvmunmap(&mut self, va: UVAddr, npages: usize, do_free: i32) {
+        if va.into_usize().wrapping_rem(PGSIZE) != 0 {
+            panic!("uvmunmap: not aligned");
+        }
+        let mut a = va.into_usize();
```
Comment (jeehoonkang) [inline kernel-rs/src/vm.rs:463]: 눈에 보인 김에.. range iterator를 쓰도록 refactoring 가능한가요?

**Review by jeehoonkang — CHANGES_REQUESTED** [review]

File: kernel-rs/src/proc.rs (hunk 1 above)
Comment (kimjungwow) [reply kernel-rs/src/proc.rs:586]: 1. `alloc()`에서 err 리턴하는 경우
  - **ProcGuard::drop()에서 `Procstate`가 `UNUSED` 인 경우 `freeproc()`을 실행하도록 리팩토링하면 좋겠습니다.**
    - `ProcGuard`를 drop할 때마다 `freeproc()`이 실행되는 것은 옳지 않다고 생각합니다.
    - `freeproc()`에서만 `Procstate`를 `UNUSED`로 바꾸며, 다른 필드들에는 `ptr::null_mut()`등 빈 값을 넣기 때문에 기존의 구현에서 큰 변화가 필요 없습니다. `alloc()` 실행 후가 아닌 `alloc()` 안에서 `ProcState`를 `RUNNABLE`로 바꿔주기만 하면 될 것 같습니다.
2. 아래 `proc_pagetable()`에서 err 리턴하는 경우
  - **`PageTable<A>::drop()`을 만들면 좋겠습니다.** 
    - `PageTable<A>::drop()`이 `uvmfree(sz)`를 호출 하면 어떨까요? 
      - 이를 위해 `ProcData`의 `sz`를 `ProcData.pagetable` 안으로 옮길 계획입니다. (`ProcData.sz`는 Size of process memory (bytes)를 의미합니다) 
    - xv6의 의도 보존을 위해 `uvmfree()`는 `vm.rs`에 계속 두면 좋겠습니다.

이 PR에서 할 수 있을 것 같지만, 새로운 이슈에서 다루는게 좋다고 생각하시면 그렇게 하겠습니다 😃 @jeehoonkang

Comment (jeehoonkang) [reply kernel-rs/src/proc.rs:586]: 다른 이슈에서 합시다.  새 이슈 만들어주세요

File: kernel-rs/src/vm.rs (hunk 4 above)
Comment (kimjungwow) [reply kernel-rs/src/vm.rs]: https://github.com/kaist-cp/rv6/pull/291/commits/49f6064ccfbb60b8d9f647b4584a88766393a55c 수정했습니다!

File: kernel-rs/src/vm.rs (hunk 3 above)
Comment (kimjungwow) [reply kernel-rs/src/vm.rs]: https://github.com/kaist-cp/rv6/pull/291/commits/3d2cbde48da28dd49911e97dfb2e4cf1cea342d5 수정했습니다!

File: kernel-rs/src/proc.rs (hunk 2 above)
Comment (kimjungwow) [reply kernel-rs/src/proc.rs:947]: https://packages.ubuntu.com/search?keywords=qemu-system-misc&searchon=names&suite=all&section=all

우분투 20.04를 기준으로 qemu-system-misc 1:4.2-3ubuntu6 보다 더 최신 버전인 6.9, 6.6에서는 부팅이 여전히 되지 않습니다. 계속 6을 사용해야 할 것 같습니다.

별도의 PR에서 `README-rv6.md`에 적어두겠습니다.

**Comment by kimjungwow** [comment]
@jeehoonkang 교수님, 리뷰 요청드립니다!! :)
- https://github.com/kaist-cp/rv6/pull/291#discussion_r530408628 는 이 PR이 머지되면 이슈 생성하도록 하겠습니다.

File: kernel-rs/src/proc.rs (hunk 1 above)
Comment (kimjungwow) [reply kernel-rs/src/proc.rs:585]: #295 에서 논의하겠습니다!

**Pull request merged**